libs/base/System/Directory.idr

```diff
@@ -97,9 +98,39 @@ removeDir : HasIO io => String -> io ()
 removeDir dirName = primIO (prim__removeDir dirName)
 
 export
-dirEntry : HasIO io => Directory -> io (Either FileError String)
-dirEntry (MkDir d)
+nextDirEntry : HasIO io => Directory -> io (Either FileError (Maybe String))
+nextDirEntry (MkDir d)
     = do res <- primIO (prim__dirEntry d)
          if prim__nullPtr res /= 0
-            then returnError
-            else ok (prim__getString res)
+            then if !(getErrno) /= 0
+                    then returnError
+                    else pure $ Right Nothing
+            else do let n = prim__getString res
+                    if n == "." || n == ".."
+                       then assert_total $ nextDirEntry (MkDir d)
```

Comment:
I'm slightly confused by this bit. Is this not an infinite loop? I read this as "if the next entry is either the current one or the parent one, loop on `MkDir d` does this work because of the side effect of `prim__dirEntry`?

Reply:
> does this work because of the side effect of prim__dirEntry?

Yes.